Compositional well

GitPaean opened the pull request:
This PR is providing some necessary elements for the compositional simulation with wells. It mostly including the following elements, 

1. parsing the schedule keywords WINJGAS and WELLSTRE
2. making PTFlash::solve() can handle arbitrary number of derivatives from the input AD variables. 
3. adding WAMF in the summary keywords

The PR can be reviewed and merged by itself without the downstream PR https://github.com/OPM/opm-simulators/pull/6042 .

GitPaean commented:
jenkins build this please

bska commented on opm/input/eclipse/Schedule/ScheduleState.hpp:
```diff
@@ -500,6 +500,9 @@ namespace Opm {
         // constant flux aquifers
         std::unordered_map<int, SingleAquiferFlux> aqufluxs;
         BCProp bcprop;
+        // injection streams for compostional STREAM injection using WINJGAS 
+        std::unordered_map<std::string, std::vector<double>> inj_streams;
```
Please use `map_member<>`.  It really is the primary solution to this problem.

GitPaean replied:
Sure.  I went back and forth on this  and thanks for the suggestion.  I will do that. 

not very familiar with the design here.  Will you explain a little bit what is the difference here for future record?

bska replied:
> not very familiar with the design here. Will you explain a little bit what is the difference here for future record?

Sure.  The point is that `map_member<K,V>` is more or less a `map<K, shared_ptr<V>>` with some additional logic on top.  This enables sharing data between `ScheduleState` objects at different report steps if there is no intervening keyword to add/change the data.

GitPaean replied:
The design looks like `V` needs to be a struct containing `name()` , which is the key .

GitPaean replied:
Is it okay that I add a member function 

```c++
void update(const K& key, std::shared_ptr<T> value);
```

or you think it will break the design.

GitPaean replied:
It looks it might work, but it will be on draft mode for some while until we figure out how things work in the opm-simulators.  I will check what Jenkins think about it. 

If it does not later, I will make a StreamComposition struct to make it the same with Well or Group to following the style change.

GitPaean commented:
jenkins build this please

GitPaean commented:
jenkins build this opm-simulators=6042 please

GitPaean commented:
jenkins build this opm-simulators=6042 please

GitPaean commented:
jenkins build this opm-simulators=6042 please

GitPaean commented:
jenkins build this opm-simulators=6042 please

GitPaean commented:
jenkins build this please

GitPaean commented:
I think this PR is ready for review and potential merging.  It can go in without the downstream PR.

akva2 commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -445,6 +445,25 @@ void handleWCYCLE(HandlerContext& handlerContext)
+
+        const double sum = std::accumulate(compostion.begin(), compostion.end(), 0.0);
+        if (std::abs(sum - 1.0) > std::numeric_limits<double>::epsilon()) {
+            std::string msg = fmt::format("The sum of the composition values for stream '{}' is not 1.0, but {}.", stream_name, sum);
```
const or eliminate temporary

akva2 commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -527,6 +546,41 @@ void handleWELOPEN(HandlerContext& handlerContext)
+        const std::string& wellNamePattern = record.getItem("WELL").getTrimmedString(0);
+        const auto well_names = HandlerContext.wellNames(wellNamePattern, false);
+        const std::string& fluid_nature = record.getItem("FLUID").getTrimmedString(0);
+        // TODO: technically, only the firt two characters are significant
```
first

akva2 commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -527,6 +546,41 @@ void handleWELOPEN(HandlerContext& handlerContext)
+        // TODO: technically, only the firt two characters are significant
+        // TODO: we need to test it out whether only the first two characters matter
+        if (fluid_nature != "STREAM") {
+            std::string msg = fmt::format("The fluid nature '{}' is not supported in WINJGAS keyword.", fluid_nature);
```
const or eliminate temporary

akva2 commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -527,6 +546,41 @@ void handleWELOPEN(HandlerContext& handlerContext)
+        // we make sure the stream is defined in WELLSTRE keyword
+        const auto& inj_streams = HandlerContext.state().inj_streams;
+        if (!inj_streams.has(stream_name)) {
+            std::string msg = fmt::format("The stream '{}' is not defined in WELLSTRE keyword.", stream_name);
```
const or eliminate temporary

akva2 commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -527,6 +546,41 @@ void handleWELOPEN(HandlerContext& handlerContext)
+        }
+
+        auto well2 = HandlerContext.state().wells.get(well_names[0]);
+        // if (well2.isProducer()) {
```
no new commented code please

akva2 commented on opm/input/eclipse/EclipseState/SummaryConfig/SummaryConfig.cpp:
```diff
@@ -394,7 +394,7 @@ struct SummaryConfigContext {
     {
         static const auto well_comp_kw = keyword_set {
-             "WXMF", "WYMF", "WZMF", "WCGMR", "WCOMR"
+             "WAMF", "WXMF", "WYMF", "WZMF", "WCGMR", "WCOMR"
```
maybe add trailing comma while at it.

akva2 commented on opm/material/constraintsolvers/PTFlash.hpp:
```diff
@@ -992,7 +992,7 @@ class PTFlash
         // pressure joins
 
-        constexpr size_t num_deri = numComponents;
+        constexpr size_t num_deri = InputEval::numVars; // numComponents;
```
just remove the trailing comment

akva2 commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -451,7 +451,11 @@ void handleWELLSTRE(HandlerContext& handlerContext)
-        // TODO: we need to check the number of the values in the composition is the same as the number of components
+        const std::size_t num_comps = handlerContext.static_schedule().m_runspec.numComps();
+        if (compostion.size() != num_comps) {
+            std::string msg = fmt::format("The number of the composition values for stream '{}' is not the same as the number of components.", stream_name);
```
const or eliminate temporary

akva2 commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -568,19 +574,21 @@ void handleWINJGAS(HandlerContext& handlerContext)
+        for (const auto& well_name : well_names) {
+            auto well2 = handlerContext.state().wells.get(well_name);
+            if (well2.isProducer()) {
+                std::string msg = fmt::format(
```
const or eliminate temporary

GitPaean commented:
The failure is due to the following message, 
```
Error: Unrecoverable errors while loading input:
Problem with keyword WINJGAS
In /var/lib/jenkins/workspace/opm-common-PR-builder/deps/opm-tests/co2store/CO2STORE_GASWAT.DATA line 186
The well 'INJ1' is a producer, not an injector, can not be specified with WINJGAS.
```

akva2 commented:
uhm, are we sure it's not just a broken test case?

GitPaean commented:
> uhm, are we sure it's not just a broken test case? it seems like a very good safety check to me even if we have to reorder those files.

Basically, the schedule parsing in OPM is in sequential order. We have multiple occasions that some DATA file is parsed by other simulator while rejected by OPM. 

The direct issue for this incident is that all the wells are created by default as producers when creating the wells based on WELSPECS, 
```c++
        Well well(wellName,
                  group,
                  timeStep,
                  0,
                  headI, headJ,
                  ref_depth,
                  WellType(preferredPhase),
```
```c++
WellType::WellType(const Phase phase)
    : WellType { true, phase }
{}
```

Then we later update whether it is a producer or injector based on WCONP* keywords or WCONJ* keywords. 

For the specific DATA file, 

the order of the keywords are 

WELSPECS
COMPDAT
WINJGAS
WCONINJE

Then it is a decision about whether we want to let this DATA pass. 
I am okay to let is pass. If it happens that the well is a producer, then WINJGAS setup is just not used. 

I am open with other suggestions also.

GitPaean commented:
> When writing new keyword handlers, we generally prefer to use compiled item names instead of string literals. Other than that this looks good to me, with a couple of opportunities to use `move()` instead of copy.

Thanks for the reviewing comments, will come back to address them later to allow other more urgent issue.

akva2 commented:
yes i understood the problem/point. my question is if  we should force sane sequential order. what is the semantics of the input file format? as far as i know, the semantics are sequential where order matters. if this is not actually the case then we indeed have to forgo the sanity check.

GitPaean commented:
I think for the problematic DATA files, the way they specify the things, I mean specifying WINJGAS before making it a injector through WCONINJE. 

```
WELSPECS
COMPDAT
WINJGAS
WCONINJE
```
we should make OPM can handle it. Otherwise, in the future we might face complaining from the user side and asking the users to modify the DATA file is not always easy. 

What we can do (not in this PR), we can have a consistence checking for each report steps when all the relevant keywords are parsed. 

For example if happens for one report step, the following situation specification happens, 
```
WELSPECS
COMPDAT
WINJGAS
WCONINJE (with or without this one)
WCONPROD
```

During the consistence checking, it shows indeed WINJGAS is specified for a producer, we should throw and gives helpful instructions for the users.

GitPaean commented on opm/input/eclipse/Schedule/Well/WellKeywordHandlers.cpp:
```diff
@@ -445,6 +445,29 @@ void handleWCYCLE(HandlerContext& handlerContext)
+{
+    auto& inj_streams = handlerContext.state().inj_streams;
+    for (const auto& record : handlerContext.keyword) {
+        const std::string& stream_name = record.getItem("STREAM").getTrimmedString(0);
+        const auto& composition = record.getItem("COMPOSITIONS").getData<double>();
```
I am using `const auto& composition` here and removing the std::move() later.  I feel more comfortable to keep the composition const.

GitPaean commented:
> > When writing new keyword handlers, we generally prefer to use compiled item names instead of string literals. Other than that this looks good to me, with a couple of opportunities to use `move()` instead of copy.
> 
> Thanks for the reviewing comments, will come back to address them later to allow other more urgent issue.

I believe I addressed the reviewing comments.

GitPaean commented:
jenkins build this please

bska commented:
>> When writing new keyword handlers, we generally prefer to use compiled item names instead of string literals
>
> I believe I addressed the reviewing comments.

You did for `WELLSTRE`, but `handleWINJGAS()` still uses string literals to identify items.

GitPaean commented:
> > > When writing new keyword handlers, we generally prefer to use compiled item names instead of string literals
> > 
> > 
> > I believe I addressed the reviewing comments.
> 
> You did for `WELLSTRE`, but `handleWINJGAS()` still uses string literals to identify items.

Oh, sorry for missing it.  now it should be updated.

GitPaean commented:
jenkins build this please

bska reviewed (approved):
Thanks a lot for the updates.  This looks good to me now and I'll merge into master.
